[processor/spanmetrics] Support for specifying Resource Attributes on metrics generated from traces

Tenaria opened the pull request:
**Description:** 
Currently there is a bug / no logic to differentiate between resource attributes. This PR adds the feature to allow users to optionally specify resource attributes to append similar to the existing dimensions mechanism. 

The logic now also indexes by the specified resource attributes to ensure accurate aggregation of data.

Service Name has been moved to become a default resource attribute instead of attribute as per the resource semantic convention https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/resource/semantic_conventions.

There is a bit of discrepancy between the use of terms "Dimensions" and "Attributes". It seems like "Attributes" is more commonly used so I have used that term in this PR for adding in resource attributes. I propose that we rename dimensions to attributes in the future, altho there will need to be some backwards compatibility added to the config to support this.

**Link to tracking Issue:** 
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/6486

**Testing:** <Describe what testing was performed and which tests were added.>
Testing added to ensure new structure/hierarchy of metrics under instrumentationLibraryMetrics structure under Resource is generated correctly.

**Documentation:**
Usage of the new config option `resource_attributes` added to README.md

Tenaria committed:
starts initial work on optional resource attributes config to span metrics processor

Tenaria committed:
adds test to check metric length when copying resources

Tenaria committed:
adds changelog and fixes tests

Tenaria committed:
Merge branch 'main' into OBC-256-Resource-Attributes-Merge-Upstream

Tenaria committed:
fix tests

Tenaria committed:
remove empty new line

Tenaria committed:
updates changelog

Tenaria committed:
add accidentally removed comment back in

chenzhihao commented on processor/spanmetricsprocessor/config.go:
```diff
@@ -20,8 +20,8 @@ import (
-// Dimension defines the dimension name and optional default value if the Dimension is missing from a span attribute.
-type Dimension struct {
+// KeyValuePair defines the key and optional default value if the key is missing from a span attribute.
+type KeyValuePair struct {
```
I think it's only key but not value? How about calling it `DimensionKey`? or `AttributeKey`

Tenaria replied:
ah yeah, I'll revert this

MovieStoreGuy commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -136,9 +149,9 @@ func mapDurationsToMillis(vs []time.Duration) []float64 {
+func validateDimensions(dimensions []KeyValuePair, defaults []string) error {
 	labelNames := make(map[string]struct{})
-	for _, key := range []string{serviceNameKey, spanKindKey, statusCodeKey} {
+	for _, key := range defaults {
 		labelNames[key] = struct{}{}
 		labelNames[sanitize(key)] = struct{}{}
 	}
```
This section of code could be optimised to increase performance by using indexing instead of a for each style loop.

```suggestion
for i := 0; i < len(defaults); i++ {
    labelNames[defaults[i]] = struct{}{}
    labelNames[sanitize(defaults[i])] = struct{}{}
}
```
The reason for this is that for each style loop creates a new copy on each iteration and we can reduce it here.

MovieStoreGuy commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -231,21 +244,40 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+
+		// if the service name doesn't exist, we treat it as invalid and do not generate a trace
+		if _, ok := resourceAttributesMap.Get(serviceNameKey); !ok {
+			continue
```
Hazzard: You can potentially have held the read lock and not freed it here, this will block any future attempts to `p.lock.Lock()`

MovieStoreGuy commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -231,21 +244,40 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
-	p.collectLatencyMetrics(ilm)
-	p.lock.RUnlock()
+		ilm := rm.InstrumentationLibraryMetrics().AppendEmpty()
+		ilm.InstrumentationLibrary().SetName("spanmetricsprocessor")
```
The string used here should be declared as a constant somewhere.

MovieStoreGuy commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -231,21 +244,40 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
-func (p *processorImp) collectLatencyMetrics(ilm pdata.InstrumentationLibraryMetrics) {
-	for key := range p.latencyCount {
+func (p *processorImp) collectLatencyMetrics(ilm pdata.InstrumentationLibraryMetrics, resAttrKey resourceKey) {
+	for mKey := range p.latencyCount[resAttrKey] {
```
I am not certain but you may have to hold the read lock for this loop since it could change underneath your feet.

Tenaria replied:
The lock is held in the code that calls this function so it should be fine right?

MovieStoreGuy commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -257,20 +289,20 @@ func (p *processorImp) collectLatencyMetrics(ilm pdata.InstrumentationLibraryMet
-func (p *processorImp) collectCallMetrics(ilm pdata.InstrumentationLibraryMetrics) {
-	for key := range p.callSum {
+func (p *processorImp) collectCallMetrics(ilm pdata.InstrumentationLibraryMetrics, resAttrKey resourceKey) {
+	for mKey := range p.callSum[resAttrKey] {
```
I think this is the same as well, you'll need to hold a lock here.

jamesmoessis replied:
Yeah this is fine

MovieStoreGuy commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -322,54 +354,82 @@ func (p *processorImp) aggregateMetricsForSpan(serviceName string, span pdata.Sp
-func (p *processorImp) updateCallMetrics(key metricKey) {
-	p.callSum[key]++
+func (p *processorImp) updateCallMetrics(resourceAttrKey resourceKey, mKey metricKey) {
+	if _, ok := p.callSum[resourceAttrKey]; ok {
```
Potential for lock as well.

Tenaria replied:
The lock is held in the code that calls this function so it should be fine right?

MovieStoreGuy replied:
I worry that someone could mistake these functions safe for concurrent access and trip up.

From what I understand, it is depends on the caller to make sure this happens?

It seems okay since the method is not public, however, it does raise a question if this should be a seperate function or just one big function?

jamesmoessis replied:
`ConsumeTraces()` should be concurrent-safe and that's it. IMO it's kind of impractical to expect every function here to be safe, and it's reasonable to expect the caller to take concurrency into account. That is how the component is designed before we touched it anyway.

chenzhihao replied:
I've put the change into this commit: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6717/commits/fd42b768ac6bd6dd1a7b541a8f9ed621002c0f48

chenzhihao commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -66,24 +67,32 @@ type processorImp struct {
+	resourceAttrList map[resourceKey]bool
 
-	// A cache of dimension key-value maps keyed by a unique identifier formed by a concatenation of its values:
+	// A cache of dimension and resource attribute key-value maps keyed by a unique identifier formed by a concatenation of its values:
```
it's just the cache of the dimensions but not resource attributes.

chenzhihao replied:
I guess we forgot to reset the comment back after you split the changed cache into two

Tenaria replied:
fixed (separated into it's own comment)

Tenaria committed:
revert variable name change and fix comment

chenzhihao commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -56,6 +56,7 @@ type exemplarData struct {
 }
 
 type metricKey string
+type resourceKey string
```
👍 yeah I think this makes the code more readable.

Tenaria committed:
revert variable name cahnge

jamesmoessis commented on CHANGELOG.md:
```diff
@@ -5,12 +5,16 @@
 ## 🛑 Breaking changes 🛑
 
 - `memcachedreceiver`: Update metric names (#6594)
+- `spanmetricproccessor`: service.name attribute added as a default resource attribute instead of attribute in generated metrics (#6717)
```
I don't think you need to update the changelog, the maintainers will do this upon a new release

Tenaria replied:
I saw in Zhihao's PR that there was a changelog added (see comment https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6503#pullrequestreview-829682349). @MovieStoreGuy can you confirm whether we update the changelog or the maintainers do?

MovieStoreGuy replied:
I think it is fine for us to update the changelog, we have the greatest overall idea of what has changed so I think it makes sense to add it. However, If a maintainer says otherwise, I am also fine with it.

Aneurysm9 replied:
We would definitely prefer that you add a changelog entry, and there is CI automation that will complain if you don't!  It generally makes it easier for us to prepare releases when we don't need to track down all the changes that have been made and think about how to describe them.

jamesmoessis commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -68,22 +69,31 @@ type processorImp struct {
-	latencyCount         map[metricKey]uint64
-	latencySum           map[metricKey]float64
-	latencyBucketCounts  map[metricKey][]uint64
+	latencyCount         map[resourceKey]map[metricKey]uint64
```
Why does this logic need to change, and why does the map structure need to become more complex? Currently it works in a way where each unique metric key is mapped to a unique attribute set. Can the metricKey just take into account the resource attributes, rather than having an entirely new key?

Tenaria replied:
This enables quicker look up time. The way OTel defines the structure to hold metrics is some thing roughly like (I've omitted the instrumentation library to keep things simplified, if you want to see the actual structure you can check out the metric proto definition):

--> Resource 1
   -> resource attributes (service.name == my.service.1, etc)
   -> metric 1
        -> data point 1
        -> data point 2
  -> metric 2
       -> data point 3
-> Resource 2
    -> resource attributes (service.name == my.service.2, etc)
    -> metric 3
      -> data point 4
etc

This means we need to index on resource key since to build that metric structure we need to find all the metrics/data points that have the same resource attributes and insert that way. (There wasn't any way to lookup on this structure afaict, it's just indexed by integers so we needed to insert all the relevant metrics/dp at the time of creation of the resource/resource attribute structure). 

If we continued to use the singular map structure that they used, we'd need to iterate over the entire metric key set and check if the key is prepended with the same resource attributes for every resource attributes - this would increase the time complexity.

jamesmoessis commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -231,21 +245,40 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+		resourceAttrKey := resourceKey(key)
+		resourceAttributesMap := p.resourceKeyToDimensions[resourceAttrKey]
+
+		// if the service name doesn't exist, we treat it as invalid and do not generate a trace
```
Would this be a breaking change to those using this processor without `service.name`?

Tenaria replied:
roughly answered here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6717#discussion_r768113596

but `service.name` is defined as a mandatory resource attribute by the [OTel semantic convention](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/README.md) so it should be there. Happy to remove this if we want, it shouldn't break any logic. The main breaking change is that we are moving the service.name field from being extracted from a resource attribute and transformed into a metric normal attribute (dimension) to being extracted from a resource attribute and transformed into a metric resource attribute as default behaviour.

jamesmoessis commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -231,21 +245,40 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+
+		rm := rms.AppendEmpty()
+
+		resourceAttributesMap.Range(func(k string, v pdata.AttributeValue) bool {
```
Can you put a comment on this line? I'm not 100% sure what it's doing

Tenaria replied:
Added in a comment (which I'll commit soon). Going to resolve this conversation so I know which comments I've addressed but if you are still unsure feel free to re-comment.

jamesmoessis commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -231,21 +245,40 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+		ilm.InstrumentationLibrary().SetName("spanmetricsprocessor")
+
+		// build metrics per resource
+		p.collectCallMetrics(ilm, resourceAttrKey)
```
I feel like we can (probably?) create new metrics under new resources without complicating the maps with `resourceAttrKey`. Happy to be shown otherwise. 

But can't we iterate through the resources and still maintain the same metric keys? They will still be unique and include the resource attributes.

It's just quite a big change and if you want this to be merged upstream then I have worries if it's big and complex, there could be some push back.

Tenaria replied:
Yeah, we could in theory create a new resource for every metric even if they have the same resource attributes but this way is more optimised for the processor in terms of memory. It's a trade-off I guess, simplicity of code vs memory. Personally I think memory optimisation is more important and I don't think the changes are that complex (albeit definitely more complex than the original). Do you think it's more worth it to have simple code?

MovieStoreGuy commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -322,54 +354,82 @@ func (p *processorImp) aggregateMetricsForSpan(serviceName string, span pdata.Sp
-	if _, ok := p.latencyExemplarsData[key]; !ok {
-		p.latencyExemplarsData[key] = []exemplarData{}
+func (p *processorImp) updateLatencyExemplars(resourceAttrKey resourceKey, mKey metricKey, value float64, traceID pdata.TraceID) {
+	if _, ok := p.latencyExemplarsData[resourceAttrKey]; ok {
+		if _, ok := p.latencyExemplarsData[resourceAttrKey][mKey]; !ok {
+			p.latencyExemplarsData[resourceAttrKey][mKey] = []exemplarData{}
+		}
+	} else {
+		p.latencyExemplarsData[resourceAttrKey] = make(map[metricKey][]exemplarData)
+		p.latencyExemplarsData[resourceAttrKey][mKey] = []exemplarData{}
```
If you flip the condition here, you can save your if else statement.

```suggestion
_, ok := p.latencyExamplarsData[p.resourceAttrKey]
if !ok {
    p.latencyExamplarsData[p.resourceAttrKey] = make(map[metricKey][]examplarData)
    p.latencyExamplarsData[p.resourceAttrKey][mkey] = make([]examplarData)
}

if _, ok = p.latencyExamplarsData[p.resourceAttrKey][mkey]; !ok {
   p.latencyExamplarsData[p.resourceAttrKey][mkey] = make([]examplarData)
}
```

Just avoids nested if statements and keeps performance consistent across conditions.

jamesmoessis commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -381,37 +442,72 @@ func (p *processorImp) buildDimensionKVs(serviceName string, span pdata.Span, op
 	return k
 }
 
+// buildResourceAttrKey builds the metric key from the service name and will attempt to add any additional resource attributes
```
Are we making the assumption that every resource revolves around service name? If so what the is reasoning behind this

Tenaria replied:
OTel defines service.name as a mandatory resource attribute (see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/README.md). Originally, the processor defined service.name as a mandatory attribute and fetched it from the resource attributes then set it as a normal attribute. This didn't make much sense to me so I altered the behaviour so that it's a default resource attribute.

jamesmoessis commented on processor/spanmetricsprocessor/processor_test.go:
```diff
@@ -523,17 +622,34 @@ func newOTLPExporters(t *testing.T) (*otlpexporter.Config, component.MetricsExpo
+	assert.Equal(t, resourceKey("a"), rk1)
 }
 
 func TestBuildKeyWithDimensions(t *testing.T) {
```
would be nice to have `t.Parallel()` here for test speed.

MovieStoreGuy reviewed (changes requested):
I need to go through another pass of this but there is a lot has changed however, I worry that access to internal data resources is not using atomic access and would ultimate lead to either incorrect data being emitted or worst case, a panic.

Could I please ask you to double check each method that reads or depends on values stored inside the process.

Tenaria committed:
adds test to ensure ordering of keys used for aggregation

Tenaria commented on processor/spanmetricsprocessor/config.go:
```diff
@@ -20,8 +20,8 @@ import (
-// Dimension defines the dimension name and optional default value if the Dimension is missing from a span attribute.
-type Dimension struct {
+// KeyValuePair defines the key and optional default value if the key is missing from a span attribute.
+type KeyValuePair struct {
```
Undid these changes, pls ignore :D

Tenaria commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -34,8 +34,8 @@ import (
 	serviceNameKey     = conventions.AttributeServiceName
-	operationKey       = "operation"   // OpenTelemetry non-standard constant.
-	spanKindKey        = "span.kind"   // OpenTelemetry non-standard constant.
+	operationKey       = "operation" // OpenTelemetry non-standard constant.
```
ah yeah my bad, I was going to change `span.kind` to use the constant defined by OTel but I didn't want to add more changes to this PR and so I manually reverted it but messed up the whitespace. I'll fix this.

Tenaria committed:
fixes lock held; optimise for loop; add parallel in test; add comments

Tenaria committed:
reverse conditions to simplify code

Tenaria committed:
update tests to use defined constant instrumentationLibraryName

albertteoh commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -381,37 +449,72 @@ func (p *processorImp) buildDimensionKVs(serviceName string, span pdata.Span, op
+	return dims
+}
+
+func concatDimensionValue(metricKeyBuilder *strings.Builder, value string) {
```
I like your suggestion of simplifying this and I propose to build on this further with the following goals in mind:
- Remove the assumption that a new/empty metricKeyBuilder should be supplied to `concatDimensionValue` in order to avoid prefixing the separator.
- Remove the assumption that `metricKeyBuilder` is not `nil`.
- Remove the need to provide a reference to `metricKeyBuilder` each time, which is a means of maintaining the key building "in progress" state.
- Make the usage simpler and more intuitive with a more idiomatic function name (as well as removing the need to pass a valid `StringBuilder`).
 
So instead of:

```
	var metricKeyBuilder strings.Builder
	concatDimensionValue(&metricKeyBuilder, span.Name())
	concatDimensionValue(&metricKeyBuilder, span.Kind().String())
	concatDimensionValue(&metricKeyBuilder, span.Status().Code().String())
```

The usage could be simplified to:
```
        mkb := keybuilder.NewMetricKeyBuilder()
	mkb.Append(serviceName)
	mkb.Append(span.Name())
	mkb.Append(span.Kind().String())
	mkb.Append(span.Status().Code().String())
```

My proposal is to separate and encapsulate the concern of metrics key building from the processor via a package `processor/spanmetricsprocessor/keybuilder`:
```
package keybuilder

import "strings"

const (
	metricKeySeparator = string(byte(0))
)

type (
	MetricKeyBuilder interface {
		Append(value string)
		String() string
	}

	metricKeyBuilder struct {
		sb        strings.Builder
		separator string
	}
)

func NewMetricKeyBuilder() MetricKeyBuilder {
	return &metricKeyBuilder{}
}

func (mkb *metricKeyBuilder) Append(value string) {
	// It's worth noting that from pprof benchmarks, WriteString is the most expensive operation of this processor.
	// Specifically, the need to grow the underlying []byte slice to make room for the appended string.
	mkb.sb.WriteString(mkb.separator)
	mkb.sb.WriteString(value)
	mkb.separator = metricKeySeparator
}

func (mkb *metricKeyBuilder) String() string {
	return mkb.sb.String()
}
```

Let me know what you think.

MovieStoreGuy replied:
I really like this approach, the only thing I would change is the method name of `keybuilder.NewMetricKeyBuilder()` to `keybuilder.New()` to avoid package name stuttering. 

I am fairly certain that `strings.Builder` has a `.Grow(by int)` method that we can use on the NewMethod which can define a package level default of what we consider to be a reasonable size byte slice?

albertteoh replied:
> the only thing I would change is the method name of keybuilder.NewMetricKeyBuilder() to keybuilder.New() to avoid package name stuttering.

> I am fairly certain that strings.Builder has a .Grow(by int) method that we can use on the NewMethod which can define a package level default of what we consider to be a reasonable size byte slice?

+1 to both suggestions.

chenzhihao committed:
Merge branch 'main' into OBC-256-Resource-Attributes-Merge-Upstream

chenzhihao committed:
update the logic of method updateCallMetrics

chenzhihao committed:
fix tests

chenzhihao committed:
add todos

chenzhihao committed:
add todos

chenzhihao committed:
Merge branch 'main' into OBC-256-Resource-Attributes-Merge-Upstream

chenzhihao committed:
fix lint

chenzhihao committed:
fix lint

chenzhihao committed:
remove resourceAttrList

chenzhihao committed:
Update "resource_attributes"

- Use LRU cache for the "resource_attributes".
- Add document/usage for the "resource_attributes" configuration.

chenzhihao committed:
fix lint

chenzhihao committed:
remove duplicate p.resetExemplarData()

chenzhihao committed:
refactor how processor reset after every ConsumeTraces process

chenzhihao committed:
move keybuilder to a package

chenzhihao committed:
update comments

chenzhihao committed:
add copyright license

jamesmoessis commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -234,13 +249,13 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
 		return err
 	}
 
+	p.reset()
```
Are you sure resetting here will not affect the metrics in `m`? If there are reference types I would want to be sure that they wouldn't be affected in the following line.

chenzhihao replied:
Yes, I am sure the `m` should not be impacted. `m` is a pointer but it's not held by the processor.

I have moved it to a defer function so that we can make the downstream components get the result a little bit faster.

jamesmoessis replied:
Actually, kind of concerned with this `reset` since it's not under any lock. There was a reason that I previously reset exemplar data only inside `buildMetrics()`. Would be nice to have this concurrency verified by tests too

chenzhihao committed:
user defer for p.reset() so that downstream components can get data quicker.

chenzhihao committed:
gofmt

chenzhihao committed:
refactor how we use p.lock

chenzhihao committed:
gofmt

jamesmoessis commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -243,12 +243,19 @@ func (p *processorImp) Capabilities() consumer.Capabilities {
 		p.reset()
+		p.lock.Unlock()
 	}()
 
```
Say there is thread (a) and (b) both calling `ConsumeTraces()`. 

1. Thread (a) obtains write lock and `aggregateMetrics()` and releases lock. 
2. Thread (b) obtains write lock and `aggregateMetrics()`, releases lock and then also goes onto `buildMetrics()`.
3. Now thread (a) calls `buildMetrics()` with the data from both (a) and (b). Specifically in the delta case this would produce incorrect metrics. 

Do you agree? Potentially buildMetrics and aggregateMetrics should be under the same lock? @MovieStoreGuy any thoughts?

chenzhihao replied:
Yeah, it also makes sense. The processor is stateful and should not be used concurrently. In practice, it should not be used concurrently either based on the design of OTEL Collector.

I would change the lock to make sure the processor can only be executed serially。

tiedotguy replied:
Regarding this, and from @MovieStoreGuy 's [comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7075#discussion_r780894362):

Locks protect data. If there is no data to protect, you don't need a lock.  If something isn't thread safe, then there's data being accessed, by definition.

If the data is "the entire `processorImp`", that's fine, but it should be documented as such, and locked as such.

Also, if you're saying something isn't thread safe, you don't need a lock.  When you declare something as not thread safe, you're explicitly pushing that concern to the caller - if they want to make it thread safe, they can add their own lock.  If you have a lock (and you use it properly!), it **is** thread safe by definition, because that's what locks give you.

Aneurysm9 replied:
> Yeah, it also makes sense. The processor is stateful and should not be used concurrently. In practice, it should not be used concurrently either based on the design of OTEL Collector.
> 
> I would change the lock to make sure the processor can only be executed serially.

I think I agree with the first statement here, that the processor should not be used concurrently, but I'm not sure about the second.  There is nothing in the design of the collector that prevents, prohibits, or even discourages concurrent execution.  Any component can spawn new goroutines at any time, and it should be anticipated that receivers will do so regularly.  While a pipeline is conceptually a serial process, there may be many instantiations of the pipeline concurrently executing and any component in a pipeline may trigger delayed or asynchronous execution of the next step in the pipeline.

I would fully expect that this processor's `ConsumeTraces()` method will be invoked concurrently and thus it should ensure that appropriate measures are taken to ensure data consistency.  If it needs to ensure a consistent view of its state without the potential for changes by other concurrent invocations, a lock should probably be held for the duration of an invocation.

jamesmoessis replied:
Thankyou @Aneurysm9 for clearing this up. Fully in favour of holding the lock for the whole invocation then. 

By the way, this PR has moved to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7075 (@Tenaria can you close this one now you're back?)

chenzhihao replied:
@tiedotguy regarding the lock, it is not protecting the `processorImp()` but to make it synchronized. I've updated the comment and hope that's more explicitly now: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7075/files#diff-78f9e7446e0e4fb474786b4b2c262486887bf27479f1f8bc598376d6c3c29b72R72.

But as we discussed [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7075#discussion_r782682242),  we probably should consider running multiple sub-processor so that each sub-processing itself is synchronized(still locked by a mutex) but these sub-processors can be executed concurrently.

chenzhihao committed:
added a set of unit tests to cover the excessive concurrent usage

chenzhihao committed:
change the usage of the lock to make sure the processor can only be executed serially

chenzhihao committed:
Merge branch 'main' into OBC-256-Resource-Attributes-Merge-Upstream

chenzhihao committed:
add comments; lint

chenzhihao commented:
As @Tenaria is away(she will be back next week), we are moving this PR to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7075 so that we can edit this PR's description and rename the branch.

Sorry for the inconvenience.

Tenaria commented:
see previous comment.
